python/cudf/cudf/_lib/scalar.pyx

```diff
@@ -92,42 +111,82 @@ cdef class Scalar:
             )
+    
+    def get_device_value(self):
+        if pd.api.types.is_string_dtype(self.dtype):
+            result = _get_py_string_from_string(self.c_value)
+        elif pd.api.types.is_numeric_dtype(self.dtype):
+            result = _get_np_scalar_from_numeric(self.c_value)
+        elif pd.api.types.is_datetime64_dtype(self.dtype):
+            result = _get_np_scalar_from_timestamp64(self.c_value)
+        elif pd.api.types.is_timedelta64_dtype(self.dtype):
+            result = _get_np_scalar_from_timedelta64(self.c_value)
+        else:
+            raise ValueError(
+                "Could not convert cudf::scalar to a Python value"
+            )
+        return result
+
 
     @property
     def dtype(self):
         """
         The NumPy dtype corresponding to the data type of the underlying
         device scalar.
         """
-        cdef libcudf_types.data_type cdtype = self.c_value.get()[0].type()
+        cdef libcudf_types.data_type cdtype = self.get_c_value()[0].type()
         return cudf_to_np_types[<underlying_type_t_type_id>(cdtype.id())]
 
     @property
     def value(self):
         """
         Returns a host copy of the underlying device scalar.
         """
-        if pd.api.types.is_string_dtype(self.dtype):
-            return _get_py_string_from_string(self.c_value)
-        elif pd.api.types.is_numeric_dtype(self.dtype):
-            return _get_np_scalar_from_numeric(self.c_value)
-        elif pd.api.types.is_datetime64_dtype(self.dtype):
-            return _get_np_scalar_from_timestamp64(self.c_value)
-        elif pd.api.types.is_timedelta64_dtype(self.dtype):
-            return _get_np_scalar_from_timedelta64(self.c_value)
+        if self._host_value_current:
+            if self.is_valid():
+                return self._host_value
+            else:
+                return cudf.NA
         else:
-            raise ValueError(
-                "Could not convert cudf::scalar to a Python value"
-            )
+            result = self.get_device_value()
+            self._host_value = result
+            self._host_value_current = True
+            return result
+
+    def _sync(self):
+        """
+        If the cache is not synched, copy either the device or host value
+        to the host or device respectively. If cache is valid, do nothing
+        """
+        if self._host_value_current and self._device_value_current:
+            return
+        elif self._host_value_current and not self._device_value_current:
+            self.set_device_value(self._host_value, self._host_dtype)
+        elif self._device_value_current and not self._host_value_current:
+            self._host_value = self.get_device_value()
+
+    
+
+    cdef scalar* get_c_value(self):
+        if not self._device_value_current:
+            self.set_device_value(self._host_value, self._host_dtype)
+            self._device_value_current = True
+        return self.c_value.get() 
```

Comment:
I think we can use the safe function we have for getting the unique_ptr here to deduplicate this code with below

Reply:
Having a few issues getting this to work locally, updates to follow.

Reply:
Connected with @shwina about how to handle this best. The issue we're running into is a combination of:

1. The fact that we can't safely return a unique pointer from the cdef class without calling `move`, which isn't safe for any function we're planning on calling more than once (the second call will fail since the pointer was already moved) 
2. The fact that we don't really want to return a raw pointer to a scalar because at that point it becomes ambiguous from the callers perspective who owns the object being pointed to and who manages it's deletion (if someone wants the raw pointer it's actually *good* that they have to reach into `the_scalar.c_value.get()`)

It looks like something a little different is required here, updates to follow.

Reply:
This has been updated. Instead of returning a unique pointer from a function (which invalidates the object) we return a small cdef class that holds the unique pointer. The larger `Scalar` object owns this wrapper object.